### PR TITLE
internal/providers/system: support .d directory for base config fragments

### PR DESCRIPTION
**Pull request by sohankunkerkar**
Fixes https://github.com/coreos/ignition/issues/1101
This change is inspired by https://github.com/coreos/fedora-coreos-config/pull/626, which allows the two distros (i.e. RHCOS and FCOS) to carry different `base.ign` files with common elements. This will also support an additional .d directory for each platform ID.

We will need to make a few changes to the `fedora-coreos-config` repo especially [here](https://github.com/coreos/fedora-coreos-config/tree/testing-devel/overlay.d/05core/usr/lib/dracut/modules.d/40ignition-conf) to split the `base.ign` config and move the `afterburn-sshkeys` into a separate fragment, and make corresponding changes to the `module-setup.sh` script.

### CODE REVIEW

**Comment by cgwalters** [comment]
I only skimmed this but the idea seems sane.  I had a comment in https://github.com/coreos/fedora-coreos-config/pull/676

File: internal/providers/system/system.go [hunk 1]
```diff
@@ -30,12 +32,26 @@ import (
...
 const (
-	baseFilename = "base.ign"
 	userFilename = "user.ign"
 )
 
 func FetchBaseConfig(logger *log.Logger) (types.Config, report.Report, error) {
-	return fetchConfig(logger, baseFilename)
+	baseDirectory := "base.d"
+	// Assuming we get only single directory which is either
+	// a `base.d` or a `base.${PLATFORM_ID}.d`.
```
Comment (arithx) [inline internal/providers/system/system.go:41]: Hmm, I wonder if the implementation might be better suited as either:

- merging both `base.d` & `base.${PLATFORM_ID}.d`

or

- copying the platform stubs into the normal `base.d`

Because we can control merge ordering via file names the individual platform stubs can always come later than the absolute base *COS bits & override as needed so we don't need to duplicate all of the normal stubs that we want into each individual platform config dir.

File: internal/providers/system/system.go [hunk 2]
```diff
@@ -35,7 +38,37 @@ const (
...
+	var fullBaseConfig types.Config
+	baseConfig, _, err := fetchConfig(logger, baseFilename)
+	if err != nil {
+		logger.Err("couldn't fetch base config: %v", err)
+		return types.Config{}, report.Report{}, err
+	}
+	fullBaseConfig = latest.Merge(fullBaseConfig, baseConfig)
+
+	// For merging base.d config fragments
+	baseDConfig, _, err := fetchBaseConfig(logger, "base.d")
```
Comment (bgilbert) [inline internal/providers/system/system.go:50]: We should be merging reports, here and below and in `fetchBaseConfig`, rather than ignoring them.  They may contain warnings.

File: internal/providers/system/system.go [hunk 3]
```diff
@@ -35,7 +38,37 @@ const (
...
+
+	contents, _ := ioutil.ReadDir(distro.SystemConfigDir())
+	for _, content := range contents {
+		if content.IsDir() {
+			// To check if a `base.$PLATFORM_ID.d` directory exists and
+			// accordingly proceed with the config merging operation.
+			if strings.Contains(content.Name(), "base") && len(strings.Split(content.Name(), ".")) == 3 && content.Size() > 0 {
+				basePlatformDConfig, _, err := fetchBaseConfig(logger, content.Name())
+				if err != nil {
+					logger.Warning("no config at %q: %v", content.Name(), err)
```
Comment (bgilbert) [inline internal/providers/system/system.go:63]: This will be the common case, so a warning seems overly aggressive.

File: internal/providers/system/system.go [hunk 4]
```diff
@@ -35,7 +38,37 @@ const (
...
+		logger.Err("couldn't fetch base config: %v", err)
+		return types.Config{}, report.Report{}, err
+	}
+	fullBaseConfig = latest.Merge(fullBaseConfig, baseConfig)
+
+	// For merging base.d config fragments
+	baseDConfig, _, err := fetchBaseConfig(logger, "base.d")
+	if err != nil {
+		logger.Warning("no config at base.d: %v", err)
+	}
```
Comment (bgilbert) [inline internal/providers/system/system.go:53]: I think it'd be clearer to merge the `baseDConfig` into `fullBaseConfig` immediately, rather than keeping track of which local variables need to be merged later.

File: internal/providers/system/system.go [hunk 5]
```diff
@@ -56,3 +89,30 @@ func fetchConfig(logger *log.Logger, filename string) (types.Config, report.Repo
...
+	} else if err != nil {
+		logger.Err("couldn't read config dir %q: %v", path, err)
+		return types.Config{}, report.Report{}, err
+	}
+	if len(configs) == 0 {
+		logger.Info("no configs at %q", path)
+		return types.Config{}, report.Report{}, nil
+	}
+	for _, config := range configs {
+		file := fmt.Sprintf("%s/%s", path, config.Name())
```
Comment (bgilbert) [inline internal/providers/system/system.go:110]: `filepath.Join()`

File: internal/providers/system/system.go [hunk 6]
```diff
@@ -56,3 +71,31 @@ func fetchConfig(logger *log.Logger, filename string) (types.Config, report.Repo
...
+		return types.Config{}, report.Report{}, err
+	}
+	if len(configs) == 0 {
+		logger.Info("no configs at %q", path)
+		return types.Config{}, report.Report{}, nil
+	}
+	for _, config := range configs {
+		intermediateConfig, intermediateReport, err := fetchConfig(logger, filepath.Join(path, config.Name()))
+		if err != nil {
+			return types.Config{}, report.Report{}, err
```
Comment (bgilbert) [inline internal/providers/system/system.go:95]: The corresponding report may be non-empty.  We should capture that in the return.  We could merge it into the base report first but I don't think that's necessary.

```suggestion
			return types.Config{}, intermediateReport, err
```

File: internal/providers/system/system.go [hunk 7]
```diff
@@ -56,3 +71,31 @@ func fetchConfig(logger *log.Logger, filename string) (types.Config, report.Repo
 	}
 	return util.ParseConfig(logger, rawConfig)
 }
+
+// fetchBaseConfig is a helper function to merge all the base config fragments inside of a particular directory.
+func fetchBaseConfig(logger *log.Logger, dir string) (types.Config, report.Report, error) {
+	var baseConfig types.Config
+	var ret report.Report
```
Comment (bgilbert) [inline internal/providers/system/system.go:78]: There are multiple return values so `ret` is ambiguous.  Let's put `report` into the variable name.

File: internal/providers/system/system.go [hunk 8]
```diff
@@ -56,3 +71,31 @@ func fetchConfig(logger *log.Logger, filename string) (types.Config, report.Repo
 	}
 	return util.ParseConfig(logger, rawConfig)
 }
+
+// fetchBaseConfig is a helper function to merge all the base config fragments inside of a particular directory.
+func fetchBaseConfig(logger *log.Logger, dir string) (types.Config, report.Report, error) {
```
Comment (bgilbert) [inline internal/providers/system/system.go:76]: Can we put "directory" into the function name?  The names of `FetchBaseConfig` and `fetchBaseConfig` don't make it clear what the distinction is.

File: internal/providers/system/system.go [hunk 9]
```diff
@@ -30,12 +32,25 @@ import (
...
 
-func FetchBaseConfig(logger *log.Logger) (types.Config, report.Report, error) {
-	return fetchConfig(logger, baseFilename)
+func FetchBaseConfig(logger *log.Logger, platformName string) (types.Config, report.Report, error) {
+	var fullBaseConfig types.Config
+	baseDConfig, baseDReport, err := fetchBaseConfig(logger, "base.d")
+	if err != nil {
+		logger.Warning("no config at base.d: %v", err)
+	}
+	fullBaseConfig = latest.Merge(fullBaseConfig, baseDConfig)
```
Comment (bgilbert) [inline internal/providers/system/system.go:44]: Actually, we don't need to merge here; we can just fetch directly into `fullBaseConfig`.

File: internal/providers/system/system.go [hunk 10]
```diff
@@ -30,12 +32,25 @@ import (
...
+
+	platformDir := fmt.Sprintf("base.%q.d", platformName)
```
Comment (bgilbert) [inline internal/providers/system/system.go:46]: Thoughts on using `filepath.Join("base.platform.d", platformName)` instead?  That way we don't have a pile of `base.X.d` directories directly inside `/usr/lib/ignition`.

File: internal/providers/system/system.go [hunk 11]
```diff
@@ -30,12 +32,25 @@ import (
...
 	userFilename = "user.ign"
 )
 
-func FetchBaseConfig(logger *log.Logger) (types.Config, report.Report, error) {
-	return fetchConfig(logger, baseFilename)
+func FetchBaseConfig(logger *log.Logger, platformName string) (types.Config, report.Report, error) {
+	var fullBaseConfig types.Config
+	baseDConfig, baseDReport, err := fetchBaseConfig(logger, "base.d")
+	if err != nil {
+		logger.Warning("no config at base.d: %v", err)
```
Comment (bgilbert) [inline internal/providers/system/system.go:42]: The callee is inconsistent about when it logs errors and when it doesn't, so this may or may not log a double message.  We should clean that up.  Also, the message here is incorrect; if there's _no_ config we'll get a `nil` error.

File: internal/providers/system/system.go (hunk 10 above)
Comment (sohankunkerkar) [reply internal/providers/system/system.go]: Yeah, it makes sense to me.

File: internal/providers/system/system.go (hunk 11 above)
Comment (sohankunkerkar) [reply internal/providers/system/system.go]: So instead of returning a customized error using `logger.Warning()`, I will simply return the `err`.

File: internal/providers/system/system.go [hunk 12]
```diff
@@ -30,12 +31,24 @@ import (
...
 const (
-	baseFilename = "base.ign"
 	userFilename = "user.ign"
 )
 
-func FetchBaseConfig(logger *log.Logger) (types.Config, report.Report, error) {
-	return fetchConfig(logger, baseFilename)
+func FetchBaseConfig(logger *log.Logger, platformName string) (types.Config, report.Report, error) {
+	var fullBaseConfig types.Config
+	baseDConfig, baseDReport, err := fetchBaseDirectoryConfig(logger, "base.d")
```
Comment (bgilbert) [inline internal/providers/system/system.go:39]: What I was suggesting was:

```go
fullBaseConfig, fullReport, err := fetchBaseDirectoryConfig(logger, "base.d")
```

**Comment by sohankunkerkar** [comment]
Local testing output:
```
[    4.989300] ignition[478]: reading system config file "/usr/lib/ignition/base.d/00-core.ign"
[    4.990706] ignition[478]: reading system config file "/usr/lib/ignition/base.d/30afterburn-sshkeys-core.ign"
[    4.992401] ignition[478]: no config dir at "/usr/lib/ignition/base.platform.d/qemu"
```

**Review by bgilbert — COMMENTED** [review]
Code looks good generally.  `tests/positive/general/preemption.go` will need to be updated.  We should also have a new automated test that checks correct merging of multiple base config and platform config fragments.

File: tests/positive/general/preemption.go [hunk 13]
```diff
@@ -37,7 +37,7 @@ func init() {
 // Ignition should apply.
 func makePreemptTest(components string) types.Test {
 	longnames := map[string]string{
-		"b": "base",
+		"b": "99-test-preemption",
```
Comment (bgilbert) [inline tests/positive/general/preemption.go:40]: Why is this change needed?

File: tests/positive/general/preemption.go [hunk 14]
```diff
@@ -75,7 +75,15 @@ func makePreemptTest(components string) types.Test {
...
+		if enabled(component) && strings.Compare(component, "b") == 0 {
+			systemFiles = append(systemFiles, types.File{
+				Node: types.Node{
+					Name:      longnames[component] + ".ign",
+					Directory: "base.d",
+				},
+				Contents: makeConfig(component),
+			})
+		} else {
 			systemFiles = append(systemFiles, types.File{
```
Comment (bgilbert) [inline tests/positive/general/preemption.go:87]: We're now doing this even if the component is disabled.

File: tests/positive/general/preemption.go [hunk 15]
```diff
@@ -75,7 +75,15 @@ func makePreemptTest(components string) types.Test {
 
 	var systemFiles []types.File
 	for _, component := range []string{"b", "u"} {
-		if enabled(component) {
+		if enabled(component) && strings.Compare(component, "b") == 0 {
```
Comment (bgilbert) [inline tests/positive/general/preemption.go:78]: `component == "b"`?

File: tests/positive/general/preemption.go [hunk 16]
```diff
@@ -75,7 +75,15 @@ func makePreemptTest(components string) types.Test {
...
+			systemFiles = append(systemFiles, types.File{
+				Node: types.Node{
+					Name:      longnames[component] + ".ign",
+					Directory: "base.d",
```
Comment (bgilbert) [inline tests/positive/general/preemption.go:82]: It'd be less repetitious to conditionalize the directory rather than the entire append operation.

File: tests/positive/general/general.go [hunk 17]
```diff
@@ -41,8 +41,40 @@ func init() {
 	register.Register(register.PositiveTest, AppendConfigWithRemoteConfigData())
 	register.Register(register.PositiveTest, VersionOnlyConfig())
 	register.Register(register.PositiveTest, EmptyUserdata())
+	register.Register(register.PositiveTest, VerifyMergingOfMultipleBaseConfigs())
 }
 
+var (
```
Comment (bgilbert) [inline tests/positive/general/general.go:47]: Because of the additional test-specific data, let's put the new test into a separate file.

File: tests/positive/general/general.go [hunk 18]
```diff
@@ -706,3 +738,72 @@ func EmptyUserdata() types.Test {
...
+			Directory: "base.d",
+		},
+		Contents: string(baseConfig2),
+	}, types.File{
+		Node: types.Node{
+			Name:      "10-plaform.ign",
+			Directory: "base.platform.d/file",
+		},
+		Contents: string(platformConfig),
+	})
```
Comment (bgilbert) [inline tests/positive/general/general.go:773]: Let's test with two platform configs also.

File: tests/positive/general/general.go [hunk 19]
```diff
@@ -706,3 +738,72 @@ func EmptyUserdata() types.Test {
 		ConfigShouldBeBad: true,
 	}
 }
+
+// VerifyMergingOfMultipleBaseConfigs checks if multiple
+// base configs and/or platform configs are merged in the
+// right order.
```
Comment (bgilbert) [inline tests/positive/general/general.go:744]: It doesn't actually test ordering because each config is disjoint.  Let's additionally set up some different file contents at the same paths to ensure later config filenames override earlier ones and platform overrides base.

File: internal/providers/system/system.go [hunk 20]
```diff
@@ -30,12 +31,24 @@ import (
...
 const (
-	baseFilename = "base.ign"
 	userFilename = "user.ign"
 )
 
-func FetchBaseConfig(logger *log.Logger) (types.Config, report.Report, error) {
-	return fetchConfig(logger, baseFilename)
+func FetchBaseConfig(logger *log.Logger, platformName string) (types.Config, report.Report, error) {
+	var fullBaseConfig types.Config
+	baseDConfig, fullReport, err := fetchBaseDirectoryConfig(logger, "base.d")
```
Comment (bgilbert) [inline internal/providers/system/system.go:39]: `baseDConfig` is still unnecessary.  Per my prior comment, you can just assign directly to `fullBaseConfig`.

File: tests/positive/general/preemption.go (hunk 13 above)
Comment (sohankunkerkar) [reply tests/positive/general/preemption.go]: There's no specific reason as such but thought of providing a number prefix to show the higher priority. I think here, it would be better to keep this as-is. Let me revert this change.

File: tests/positive/general/preemption.go (hunk 14 above)
Comment (sohankunkerkar) [reply tests/positive/general/preemption.go]: Agreed! I've addressed this issue in the subsequent push.

File: tests/positive/general/general.go (hunk 17 above)
Comment (sohankunkerkar) [reply tests/positive/general/general.go]: Sure, let me fix it

File: internal/providers/system/system.go [hunk 21]
```diff
@@ -30,12 +31,24 @@ import (
...
 
 const (
-	baseFilename = "base.ign"
 	userFilename = "user.ign"
 )
 
-func FetchBaseConfig(logger *log.Logger) (types.Config, report.Report, error) {
-	return fetchConfig(logger, baseFilename)
+func FetchBaseConfig(logger *log.Logger, platformName string) (types.Config, report.Report, error) {
+	var fullBaseConfig types.Config
```
Comment (arithx) [inline internal/providers/system/system.go:38]: Is this declaration necessary?

File: internal/providers/system/system.go [hunk 22]
```diff
@@ -30,12 +31,24 @@ import (
 )
 
 const (
-	baseFilename = "base.ign"
 	userFilename = "user.ign"
 )
 
-func FetchBaseConfig(logger *log.Logger) (types.Config, report.Report, error) {
-	return fetchConfig(logger, baseFilename)
+func FetchBaseConfig(logger *log.Logger, platformName string) (types.Config, report.Report, error) {
```
Comment (arithx) [inline internal/providers/system/system.go:37]: Would be nice to add a doc string to what is happening here (and the expected/checked paths).

File: tests/positive/general/baseconfigsmerging.go [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+			Contents: "base config\n",
+		},
+	})
+	out[0].Partitions.AddFiles("ROOT", []types.File{
+		{
+			Node: types.Node{
+				Name:      "bar2",
+				Directory: "foo",
+			},
+			Contents: "platform config3\n",
```
Comment (bgilbert) [inline tests/positive/general/baseconfigsmerging.go:149]: This proves that platform overrides base, but not that base configs override each other in the right order.

File: tests/positive/general/preemption.go [hunk 24]
```diff
@@ -74,14 +74,26 @@ func makePreemptTest(components string) types.Test {
...
 	for _, component := range []string{"b", "u"} {
 		if enabled(component) {
-			systemFiles = append(systemFiles, types.File{
-				Node: types.Node{
-					Name: longnames[component] + ".ign",
-				},
-				Contents: makeConfig(component),
-			})
+			if component == "b" {
+				systemFile = types.File{
```
Comment (bgilbert) [inline tests/positive/general/preemption.go:81]: This is still more redundant than necessary.  You can just conditionalize a variable holding the field value for `Directory`.

**Review by arithx — APPROVED** [review]
WFM

File: tests/positive/general/preemption.go [hunk 25]
```diff
@@ -75,10 +75,15 @@ func makePreemptTest(components string) types.Test {
 
 	var systemFiles []types.File
 	for _, component := range []string{"b", "u"} {
+		var dir string
```
Comment (bgilbert) [inline tests/positive/general/preemption.go:78]: This could be moved one scope further in, inside the `if enabled(component)` block.

**Review by bgilbert — APPROVED** [review]
:tada:
